### ISSUE TITLE
Integration with surveyRcpp

### ISSUE THREAD

**Issue by gergness** [gergness/srvyr#9]
Preliminary notes on integrating with surveyRcpp

## Integration
I think that there could be some advantages to keeping the two packages separate, with srvyr using surveyRcpp as a dependency. Mainly, srvyr could continue to use the survey functions for things that haven't been implemented yet, and could allow users to choose between which system they use (perhaps surveyRcpp for exploratory code because it's faster, and use survey in production because it is more tested). 

What do you think @carlganz ?

## Class system
Currently surveyRcpp uses references class objects, while both dplyr and the original survey package use S3 objects. I don't think it is possible to do S3 method dispatch on RC objects, so for example, adding a method using `summarise_.surveyRcpp()` in an analagous way to `summarise_.tbl_svy()` doesn't work. (Receive error: `no applicable method for 'summarise_' applied to an object of class "c('Rcpp_svrepdesign', 'C++Object', 'envRefClass', '.environment', 'refClass', 'environment', 'refObject')"`)

I can see 3 solutions (and there may be more):
1. srvyr could make an S3 wrapper class around the surveyRcpp object, which would be a list with a single object (a surveyRcpp object). This seems like wasteful code, but should work.
2. surveyRcpp could use S3 objects. (Is this feasible @carlganz ?)
3. surveyRcpp (or srvyr) could ad RC methods for the dplyr functions such as `summarise_`. This may not be possible, because I don't know if R would look at the RC methods first, or would try to do S3. It also seems like it would make code less readable because the method definitions would be in multiple places.

## function of vector (as opposed to formula)
surveyRcpp doesn't currently support passing in vectors to the `svymean` function, but srvyr currently depends on this (at least for ungrouped data). For example:
```
chis.svy <- svrepdesign(variables = child[, !grepl("rakedw*", names(child))],
                                  repweights=child[, grepl("rakedw*", names(child))][,-1],
                                  weights=child$rakedw0, type="other", scale=1, rscales=1 )

# Works
svymean(~insany, chis.svy)

# Does not work
svymean(chis.svy$variables$insany, chis.svy)
> #  Error: expecting a single value
```
The code could be changed in srvyr, but it would create code complexity to allow for the full set of dplyr functionality (for example, calculating the mean of a newly created variable such as `svy %>% summarize(x = a - b)`. Further, the current survey code is slightly faster when using vectors.

**Comment by carlganz** [comment]
Hey @gergness ,
Hope all is well with you. Keeping `srvyr` and `surveyRcpp` separate definitely makes sense. I have been pretty bogged down with finals so I have't worked on `surveyRcpp` in a while, but I will have some time soon.

When I started working on `surveyRcpp` I thought about a few different approaches, and one of them was recreating the `survey` objects as classes in C++. I went this way partly as an exercise in learning about classes. I discovered that classes in C++ serve a different purpose than classes in R (at least when using C++ to interface with R), and as you mention above, the C++ methods don't play nicely with R. I am not going to pursue this approach.

My original plan was to fork the `survey` package, and go through it, and try to substitute C++ in different areas, but I had a lot of trouble with that. The `survey` package is sort of old-school; it uses requires/depends instead of imports (and doesn't pass a Travis as is), so I wasn't able to add `Rcpp` without breaking the package builder. Also, the `survey` package is huge, so I don't think this approach is practical.

The approach that I will likely go with is to simply write functions in C++ that use the inputs of the `survey` object as parameters. This should work with `srvyr` much nicer, and building in tests to make sure it gives the same answers as the `survey` package should be fairly easy. 

Kind Regards
